Storage/REST-API.md

```diff
@@ -0,0 +1,183 @@
+
+Successful `POST` results in either `200 (OK)` error code if the resource has been updated, or `201 (Created)`.
+
+* **DELETE**: Deletes the resource.
+
+   Similarly to `PATCH` and `PUT`, `DELETE` should be called on the URL that has been (or could have been) returned as the top-level `"url":"..."` field of the payload of `GET`.
+   
+   `DELETE` follows the same conflict resolution rules as `PATCH` and `PUT`.
+   
+`TODO: User-defined data integrity checks, and what errors are returned if they fail?`
+
+
+### Discoverability
+
+The entry point (usually, `"/"`) URL will contain the list of inner `"url_*"`-s to access respective fields ("tables") of the storage.
+
+The entry point URL may also contain the URLs by which the user can access the respective schema files. This is applicable to F# and C++.
+
+[ TO DISCUSS ] The payload when `GET`-ting the resource corresponding to an individual record (ex. `/user/123`) will contain the `{...,"url_collection":".../users"}` reference to the URL to browse the whole collection, the record from which is being requested.
+
+As a record is successfully created or modified, the response payload will contain its URL as the top-level `"url"` field.
+
+When a resource can not be found, the error payload will contain the URL to browse the respective collection.
+
+When the operation fails due to a conflict, the top-level `"url"` field will be set as well.
+
+### Pagination
+
+When retrieving the collection of records, they are returned in pages. If there are more records to browse, the top-level response object will contain the `"url_next_page"` field, as well as the `"url_previous_page"` one for the pages past the first one.
+
+As mentioned above, in the pagination mode:
+* the `"data":[...]` top-level field becomes an array, not object, and
+* the elements of this array default to the `&format=preview` format, if defined by the user.
+
+With respect to pagination:
+* The user SHOULD NOT attempt to skip through records (ex. browse page 1 with records 1..20, and then skip immediately to records 101..120.
+* The user CAN expect the records they have already seen as part of this session will not change.
+* The user SHOULD NOT expect any record they have seen or may see will exist and/or be available to them by the time the user decides to perform actions on that record.
+
+In other words, the behavior of pagination for the user mimics the architecture where on the first `GET` request to browse the collection, the complete list of possible IDs is populated on the server, and the user is just browsing through these IDs.
+
+The server will honor the entries the user has already browsed through, even if the user requests the previous page, and some entries from it have been deleted. At the same time, while browsing further, the server may re-fetch the records and re-apply the filters.
```

Comment:
Does the server guarantee that the records that the user has seen on some pages won't appear on the other pages?